### ISSUE TITLE
 Tensorflow lite model gives very different accuracy value compared to python model

### ISSUE THREAD

**Issue by jimenisimo** [tensorflow/tensorflow#21921]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.10
- **Python version**:3.6
- **Bazel version (if compiling from source)**: 0.16.0
- **CUDA/cuDNN version**:N/A
- **GPU model and memory**:N/A
- **Exact command to reproduce**: See Source Code

### Describe the problem
I am using tensorflow 1.10 Python 3.6

My code is based in the premade iris classification model provided by TensorFlow. This means, I am using a Tensorflow DNN premade classifier, with the difference following difference:

10 features instead 4.
5 classes instead 3.
The test and training files can be downloaded from the following link: https://www.dropbox.com/sh/nmu8i2i8xe6hvfq/AADQEOIHH8e-kUHQf8zmmDMDa?dl=0

I have made a code to export this classifier to a tflite format, however the accuracy in the python model is higher than 75% but when exported the accuracy decrease approximately to 45% this means approximately 30% Accuracy is lost (This is too much). I have tried the code with different set of data and in all of them the accuracy after exporting decrease a lot! This made me think that something is going wrong with the TocoConverter function or that maybe I am exporting to tflite incorrectly, missing a parameter or something like that.

I share the code in which I calculate also the accuracy of the .tflite file.

I hope some of you can identify the error, or give a possible solution

### Source code / logs

```
import argparse
import tensorflow as tf

import pandas as pd
import csv

from tensorflow.python.tools import freeze_graph
from tensorflow.python.tools import optimize_for_inference_lib
import numpy as np


parser = argparse.ArgumentParser()
parser.add_argument('--batch_size', default=100, type=int, help='batch size')
parser.add_argument('--train_steps', default=1000, type=int,
                    help='number of training steps')

features_global = None
feature_spec = None

MODEL_NAME = 'myModel'

def load_data(train_path, test_path):
    """Returns the iris dataset as (train_x, train_y), (test_x, test_y)."""

    with open(train_path, newline='') as f:
        reader = csv.reader(f)
        column_names = next(reader)

    y_name = column_names[-1]

    train = pd.read_csv(train_path, names=column_names, header=0)
    train_x, train_y = train, train.pop(y_name)

    test = pd.read_csv(test_path, names=column_names, header=0)
    test_x, test_y = test, test.pop(y_name)

    return (train_x, train_y), (test_x, test_y)


def train_input_fn(features, labels, batch_size):
    """An input function for training"""
    # Convert the inputs to a Dataset.
    dataset = tf.data.Dataset.from_tensor_slices((dict(features), labels))

    # Shuffle, repeat, and batch the examples.
    dataset = dataset.shuffle(1000).repeat().batch(batch_size)

    # Return the dataset.
    return dataset


def eval_input_fn(features, labels, batch_size):
    """An input function for evaluation or prediction"""
    features=dict(features)
    if labels is None:
        # No labels, use only features.
        inputs = features
    else:
        inputs = (features, labels)

    # Convert the inputs to a Dataset.
    dataset = tf.data.Dataset.from_tensor_slices(inputs)

    # Batch the examples
    assert batch_size is not None, "batch_size must not be None"
    dataset = dataset.batch(batch_size)

    # Return the dataset.
    return dataset


def main(argv):
    args = parser.parse_args(argv[1:])

    train_path = "trainData.csv"
    test_path = "testData.csv"

    # Fetch the data
    (train_x, train_y), (test_x, test_y) = load_data(train_path, test_path)

    # Load labels
    num_labels = 5

    # Feature columns describe how to use the input.
    my_feature_columns = []
    for key in train_x.keys():
        my_feature_columns.append(tf.feature_column.numeric_column(key=key))

    # Build 2 hidden layer DNN
    classifier = tf.estimator.DNNClassifier(
        feature_columns=my_feature_columns,
        # Two hidden layers of 10 nodes each.
        hidden_units=[100, 500],
        # The model must choose between 'num_labels' classes.
        optimizer=tf.train.AdagradOptimizer(learning_rate=0.003),
        n_classes=num_labels,
        model_dir="myModel")

    # Train the Model
    classifier.train(
        input_fn=lambda:train_input_fn(train_x, train_y,
                                                args.batch_size),
        steps=args.train_steps)

    # Evaluate the model.
    eval_result = classifier.evaluate(
        input_fn=lambda:eval_input_fn(test_x, test_y,
                                                args.batch_size))

    print('\nTest set accuracy: {accuracy:0.3f}\n'.format(**eval_result))

    # Export model
    feature_spec = tf.feature_column.make_parse_example_spec(my_feature_columns)
    serve_input_fun = tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)
    saved_model_path = classifier.export_savedmodel(
            export_dir_base="out",
            serving_input_receiver_fn=serve_input_fun,
            as_text=True,
            checkpoint_path=classifier.latest_checkpoint(),
        )
    tf.reset_default_graph()
    var = tf.Variable(0)
    with tf.Session() as sess:
        # First let's load meta graph and restore weights
        sess.run(tf.global_variables_initializer())
        latest_checkpoint_path = classifier.latest_checkpoint()
        saver = tf.train.import_meta_graph(latest_checkpoint_path + '.meta')
        saver.restore(sess, latest_checkpoint_path)

        input_arrays = ["dnn/input_from_feature_columns/input_layer/concat"]
        output_arrays = ["dnn/logits/BiasAdd"]

        frozen_graph_def = tf.graph_util.convert_variables_to_constants(
            sess, sess.graph_def,
            output_node_names=["dnn/logits/BiasAdd"])

        frozen_graph = "out/frozen_graph.pb"

        with tf.gfile.FastGFile(frozen_graph, "wb") as f:
                f.write(frozen_graph_def.SerializeToString())

        # save original graphdef to text file
        with open("estimator_graph.pbtxt", "w") as fp:
            fp.write(str(sess.graph_def))
        # save frozen graph def to text file
        with open("estimator_frozen_graph.pbtxt", "w") as fp:
            fp.write(str(frozen_graph_def))

        input_node_names = input_arrays
        output_node_name = output_arrays
        output_graph_def = optimize_for_inference_lib.optimize_for_inference(
                frozen_graph_def, input_node_names, output_node_name,
                tf.float32.as_datatype_enum)

        final_model_path = 'out/opt_' + MODEL_NAME + '.pb'
        with tf.gfile.FastGFile(final_model_path, "wb") as f:
            f.write(output_graph_def.SerializeToString())

        tflite_file = "out/iris.tflite"

        converter = tf.contrib.lite.TocoConverter.from_frozen_graph(final_model_path, input_arrays, output_arrays, input_shapes={"dnn/input_from_feature_columns/input_layer/concat": [1, 10]})
        tflite_model = converter.convert()
        open(tflite_file, "wb").write(tflite_model)

        interpreter = tf.contrib.lite.Interpreter(model_path=tflite_file)
        interpreter.allocate_tensors()

        # Get input and output tensors.
        input_details = interpreter.get_input_details()
        output_details = interpreter.get_output_details()

        # Test model on random input data.
        input_shape = input_details[0]['shape']
        # change the following line to feed into your own data.
        input_data = np.array(np.random.random_sample(input_shape), dtype=np.float32)
        resultlist = list()
        df = pd.read_csv(test_path)
        expected = df.iloc[:, -1].values.tolist()
        with open(test_path, newline='') as f:
            reader = csv.reader(f)
            column_names = next(reader)
            for x in range(1, len(expected)):
                linea = next(reader)
                linea = linea[:len(linea) - 1]
                input_data2 = np.array(linea, dtype=np.float32)
                interpreter.set_tensor(input_details[0]['index'], [input_data2])
                interpreter.invoke()
                output_data = interpreter.get_tensor(output_details[0]['index'])
                #print(output_data)
                max = 0;
                longitud = len(output_data[0])

                for k in range(0, longitud):
                    if (output_data[0][k] > output_data[0][max]):
                        max = k
                resultlist.append(max)
            print(resultlist)

        coincidences = 0
        for pred_dict, expec in zip(resultlist, expected):
            if pred_dict == expec:
                coincidences = coincidences + 1

        print("tflite Accuracy: " + str(coincidences / len(expected)))


if __name__ == '__main__':
    tf.logging.set_verbosity(tf.logging.INFO)
    tf.app.run(main)
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by jimenisimo** [comment]
Fields Updated

**Comment by YijinLiu** [comment]
I also noticed that tflite produces worse results. I am not sure it's that huge (75 VS 45) though.
The commands I used to convert to tflite are:

tf_models/research/object_detection/export_tflite_ssd_graph.py \
    --pipeline_config_path=data/gen/coco/ssdlite_mobilenet_v2.config \
    --trained_checkpoint_prefix=data/gen/coco/ssdlite_mobilenet_v2/model.ckpt-66366 \
    --output_directory=data/gen/coco/ssdlite_mobilenet_v2_exported --add_postprocessing_op=true

bazel run --config=opt tensorflow/contrib/lite/toco:toco -- --input_file=/home/chao/projects/tf-object-detect/train/data/gen/coco/ssdlite_mobilenet_v2_exported/tflite_graph.pb --output_file=/home/chao/projects/tf-object-detect/train/data/gen/coco/ssdlite_mobilenet_v2_exported/detect.tflite --input_shapes=1,300,300,3 --input_arrays=normalized_input_image_tensor --output_arrays=TFLite_Detection_PostProcess,TFLite_Detection_PostProcess:1,TFLite_Detection_PostProcess:2,TFLite_Detection_PostProcess:3 --inference_type=FLOAT --allow_custom_ops

**Comment by hequn** [comment]
@YijinLiu I am encountered with it, too.  I selected ssdlite_mobilenet_v2.config  to train and eval the model, then converted the ckpt to pb and pb to tflite step by step according to the doc, but at last it gave the worse results compared to the pb that was generated by export_inference_graph.py
python 3.6.6
tensorflow 1.10.0
bazel 0.15.2

**Comment by hequn** [comment]
Hi, guys @YijinLiu @jimenisimo I solved my problem by adding the std and mean to the input data though I did not give the  "--mean_values=128     --std_values=128" when using bazel compile.
When inferring the image, 'input_data = np.expand_dims((input_data - args.input_mean) / args.input_std, axis=0)' should be given , and the input_mean is 128/ std is 128.

**Comment by jdduke** [comment]
+Suharsh for additional guidance.

**Comment by jimenisimo** [comment]
@hequn I have no idea how to implement your solution in my code, could you help me, have you tried my code?

**Comment by hequn** [comment]
@jimenisimo These codes mentioned above are not my codes, they are in the research module 'object_detection', you may go into that and check the source codes. 
Sorry, I haven't tried you code, but it seems the acc from 75% to 45% is not similar with the situation in my test.

**Comment by alanchiao** [comment]
Once resolved, see [jimenisimo's stackoverflow post](https://stackoverflow.com/questions/52057552/tensorflow-lite-model-gives-very-different-accuracy-value-compared-to-python-mod ) for closing also. There are a number of people who have upvoted it.

**Comment by YijinLiu** [comment]
It seems to me that tflite is not able to detect objects overlap with other objects. I hope this is a bug instead of a fundamental problem of tensorflow lite's object detection postprocess implementation.

**Comment by YijinLiu** [comment]
After playing with detection_postprocess.cc a bit, I don't think these missing overlapped objects are caused by Detection_PostProcess. They don't even exist in Detection_PostProcess's input.

**Comment by alanchiao** [comment]
@YijinLiu : @achowdhery is most familiar with export_ssd script and the conversion process. 

Can we move this discussion to a separate thread since it's unrelated to the issue that @jimenisimo has?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @achowdhery: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tofulawrence** [comment]
@YijinLiu Please file a new issue. Thanks.

@jimenisimo: Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by nekulkarni** [comment]
@suharshs  @achowdhery Has there been any update about this issue? I see the same issue with tensorflow 1.10. The accuracy drops significantly (> 10%) when a frozen graph (50k steps with quantized-aware training kicking in at 48k steps) is converted to tflite using toco. The model is ssd_mobilenetv2 and it was trained with TF Object Detection API.

**Comment by ameerusman** [comment]
Apologies for bumping an old thread but wanted to ask that with recent improvements in the model, has the accuracy been improved (like it has been near 4.5 years) and can it be reliably used on a edge microcontroller or even a small Linux processor?
